gax/backoff_policy.h

```diff
@@ -0,0 +1,168 @@
+// Copyright 2019 Google Inc.  All rights reserved
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef GOOGLE_GAX_BACKOFF_POLICY_H_
+#define GOOGLE_GAX_BACKOFF_POLICY_H_
+
+#include <chrono>
+#include <memory>
+#include <random>
+
+#include "googletest/include/gtest/gtest_prod.h"
+
+namespace google {
+namespace gax {
+
+/**
+ * Define the interface for backoff policies.
+ *
+ * The client libraries need to hide partial and temporary failures from the
+ * application. Exponential backoff is generally considered a best practice when
+ * retrying operations. However, the details of how exponetial backoff is
+ * implemented and tuned varies widely. We need to give the users enough
+ * flexibility, and also provide sensible default implementations.
+ *
+ * The client library receives an object of this type and clones a new instance
+ * for each operation. That is, the application provides the library with a
+ * [Prototype](https://en.wikipedia.org/wiki/Prototype_pattern) of the policy
+ * that will be applied to each operation.
+ *
+ * [Truncated Exponential
+ * Backoff](https://cloud.google.com/storage/docs/exponential-backoff) in the
+ * Google Cloud Storage documentation.
+ *
+ */
+class BackoffPolicy {
+ public:
+  virtual ~BackoffPolicy() = default;
+
+  /**
+   * Handle an operation completion.
+   *
+   * This function is called when an operation has failed and needs to be retried.
+   * The decision to retry or not is handled by other policies.
+   *
+   * @return the delay to wait before the next retry attempt.
+   */
+  virtual std::chrono::microseconds OnCompletion() = 0;
+
+  /**
+   * Return a new copy of this object.
+   */
+  virtual std::unique_ptr<BackoffPolicy> clone() const = 0;
+
+  // Used in testing
+ protected:
+  FRIEND_TEST(ExponentialBackoffPolicy, Clone);
+  virtual std::chrono::microseconds CurrentDelay() const = 0;
```

Comment:
Thinking more on this, I really do think it is worth moving these out of the public interface. Otherwise we force other implementations to implement these methods when they may be unnecessary. See my comment below on the clone method for a suggestion on how to test it without using these methods.

Reply:
Yeah, I agree. It made sense at the time, but it's really a test-only implementation detail. Will remove.